File: src/services/parallelProcessor.ts
import { v4 as uuidv4 } from 'uuid';

export interface ProcessingBatch<T> {
  batchId: string;
  items: T[];
  batchSize: number;
  createdAt: number;
}

export interface ProcessingResult<R> {
  itemId: string;
  success: boolean;
  result?: R;
  error?: string;
  processingTime: number;
}

export interface BatchProcessingResult<R> {
  totalProcessed: number;
  successful: number;
  failed: number;
  errors: Array<{ itemId: string; error: string }>;
  processingTime: number;
  batchId: string;
  results: R[];
}

export interface ProcessingStats {
  totalProcessed: number;
  totalBatches: number;
  averageBatchTime: number;
  errorRate: number;
  throughputPerSecond: number;
}

export class ParallelProcessor<T, R> {
  private maxWorkers: number;
  private batchSize: number;
  private maxConcurrentBatches: number;
  private timeoutMs: number;
  private stats: ProcessingStats;

  constructor(options: {
    maxWorkers?: number;
    batchSize?: number;
    maxConcurrentBatches?: number;
    timeoutMs?: number;
  } = {}) {
    this.maxWorkers = options.maxWorkers || 4;
    this.batchSize = options.batchSize || 10;
    this.maxConcurrentBatches = options.maxConcurrentBatches || 3;
    this.timeoutMs = options.timeoutMs || 30000; // 30 seconds

    this.stats = {
      totalProcessed: 0,
      totalBatches: 0,
      averageBatchTime: 0,
      errorRate: 0,
      throughputPerSecond: 0,
    };
  }

  /**
   * Create batches from items for parallel processing
   */
  createBatches(items: T[], customBatchSize?: number): ProcessingBatch<T>[] {
    const effectiveBatchSize = customBatchSize || this.batchSize;
    const batches: ProcessingBatch<T>[] = [];

    // Adaptive batch sizing
    let adaptedBatchSize = effectiveBatchSize;
    if (items.length < effectiveBatchSize) {
      adaptedBatchSize = Math.max(1, Math.ceil(items.length / 2));
    } else if (items.length > effectiveBatchSize * 10) {
      adaptedBatchSize = Math.min(effectiveBatchSize * 2, Math.ceil(items.length / this.maxWorkers));
    }

    for (let i = 0; i < items.length; i += adaptedBatchSize) {
      const batchItems = items.slice(i, i + adaptedBatchSize);
      batches.push({
        batchId: uuidv4(),
        items: batchItems,
        batchSize: batchItems.length,
        createdAt: Date.now(),
      });
    }

    return batches;
  }

  /**
   * Process a single batch with concurrency control
   */
  private async processBatch<T, R>(
    batch: ProcessingBatch<T>,
    processorFn: (item: T) => Promise<R>,
    getItemId: (item: T) => string = (item) => String(item)
  ): Promise<ProcessingResult<R>[]> {
    const startTime = Date.now();
    const results: ProcessingResult<R>[] = [];

    // Create a semaphore to limit concurrent operations within the batch
    const semaphore = new Semaphore(this.maxWorkers);

    const processItem = async (item: T): Promise<ProcessingResult<R>> => {
      const itemStartTime = Date.now();
      const itemId = getItemId(item);

      try {
        await semaphore.acquire();
        
        // Add timeout to individual item processing
        const result = await Promise.race([
          processorFn(item),
          new Promise<never>((_, reject) => 
            setTimeout(() => reject(new Error('Processing timeout')), this.timeoutMs)
          )
        ]);

        return {
          itemId,
          success: true,
          result,
          processingTime: Date.now() - itemStartTime,
        };
      } catch (error) {
        return {
          itemId,
          success: false,
          error: error instanceof Error ? error.message : String(error),
          processingTime: Date.now() - itemStartTime,
        };
      } finally {
        semaphore.release();
      }
    };

    try {
      // Process all items in the batch concurrently
      const itemPromises = batch.items.map(processItem);
      const itemResults = await Promise.allSettled(itemPromises);

      // Handle settled promises
      itemResults.forEach((settledResult, index) => {
        if (settledResult.status === 'fulfilled') {
          results.push(settledResult.value);
        } else {
          results.push({
            itemId: getItemId(batch.items[index]),
            success: false,
            error: settledResult.reason?.message || 'Unknown error',
            processingTime: 0,
          });
        }
      });

      const processingTime = Date.now() - startTime;
      console.log(`Batch ${batch.batchId} processed in ${processingTime}ms`);

      return results;
    } catch (error) {
      console.error(`Critical error processing batch ${batch.batchId}:`, error);
      
      // Return error results for all items in batch
      return batch.items.map(item => ({
        itemId: getItemId(item),
        success: false,
        error: `Batch processing failed: ${error instanceof Error ? error.message : String(error)}`,
        processingTime: 0,
      }));
    }
  }

  /**
   * Process items in parallel with intelligent batching
   */
  async processParallel(
    items: T[],
    processorFn: (item: T) => Promise<R>,
    options: {
      customBatchSize?: number;
      getItemId?: (item: T) => string;
    } = {}
  ): Promise<BatchProcessingResult<R>> {
    if (items.length === 0) {
      return {
        totalProcessed: 0,
        successful: 0,
        failed: 0,
        errors: [],
        processingTime: 0,
        batchId: uuidv4(),
        results: [],
      };
    }

    const startTime = Date.now();
    const batches = this.createBatches(items, options.customBatchSize);
    const allResults: ProcessingResult<R>[] = [];

    // Process batches with concurrency limit
    const batchSemaphore = new Semaphore(this.maxConcurrentBatches);
    
    const processBatchWithSemaphore = async (batch: ProcessingBatch<T>) => {
      try {
        await batchSemaphore.acquire();
        return await this.processBatch(batch, processorFn, options.getItemId);
      } finally {
        batchSemaphore.release();
      }
    };

    try {
      // Process all batches
      const batchPromises = batches.map(processBatchWithSemaphore);
      const batchResults = await Promise.allSettled(batchPromises);

      // Flatten results
      batchResults.forEach((settledResult) => {
        if (settledResult.status === 'fulfilled') {
          allResults.push(...settledResult.value);
        } else {
          console.error('Batch processing failed:', settledResult.reason);
        }
      });

      // Calculate statistics
      const totalProcessed = allResults.length;
      const successful = allResults.filter(r => r.success).length;
      const failed = totalProcessed - successful;
      const processingTime = Date.now() - startTime;

      // Update internal stats
      this.updateStats(totalProcessed, batches.length, processingTime, failed / totalProcessed);

      // Collect successful results and errors
      const successfulResults = allResults
        .filter(r => r.success && r.result !== undefined)
        .map(r => r.result as R);

      const errors = allResults
        .filter(r => !r.success)
        .map(r => ({ itemId: r.itemId, error: r.error || 'Unknown error' }));

      return {
        totalProcessed,
        successful,
        failed,
        errors,
        processingTime,
        batchId: uuidv4(),
        results: successfulResults,
      };
    } catch (error) {
      console.error('Critical error in parallel processing:', error);
      throw error;
    }
  }

  /**
   * Process items with retry logic
   */
  async processWithRetry(
    items: T[],
    processorFn: (item: T) => Promise<R>,
    options: {
      maxRetries?: number;
      retryDelay?: number;
      customBatchSize?: number;
      getItemId?: (item: T) => string;
    } = {}
  ): Promise<BatchProcessingResult<R>> {
    const maxRetries = options.maxRetries || 3;
    const retryDelay = options.retryDelay || 1000;
    
    let currentItems = [...items];
    let allSuccessfulResults: R[] = [];
    let finalErrors: Array<{ itemId: string; error: string }> = [];
    let totalProcessingTime = 0;

    for (let attempt = 0; attempt <= maxRetries && currentItems.length > 0; attempt++) {
      const result = await this.processParallel(currentItems, processorFn, options);
      
      totalProcessingTime += result.processingTime;
      allSuccessfulResults.push(...result.results);

      if (attempt === maxRetries) {
        // Final attempt, collect remaining errors
        finalErrors.push(...result.errors);
      } else if (result.failed > 0) {
        // Prepare failed items for retry
        const failedItemIds = new Set(result.errors.map(e => e.itemId));
        currentItems = currentItems.filter(item => 
          failedItemIds.has(options.getItemId ? options.getItemId(item) : String(item))
        );

        if (currentItems.length > 0) {
          console.log(`Retrying ${currentItems.length} failed items (attempt ${attempt + 1}/${maxRetries})`);
          await new Promise(resolve => setTimeout(resolve, retryDelay * (attempt + 1)));
        }
      } else {
        // All items processed successfully
        break;
      }
    }

    return {
      totalProcessed: items.length,
      successful: allSuccessfulResults.length,
      failed: finalErrors.length,
      errors: finalErrors,
      processingTime: totalProcessingTime,
      batchId: uuidv4(),
      results: allSuccessfulResults,
    };
  }

  /**
   * Update processing statistics
   */
  private updateStats(processed: number, batches: number, timeTaken: number, errorRate: number): void {
    this.stats.totalProcessed += processed;
    this.stats.totalBatches += batches;

    // Update average batch time
    if (this.stats.totalBatches > 0) {
      const totalTime = this.stats.averageBatchTime * (this.stats.totalBatches - batches) + timeTaken;
      this.stats.averageBatchTime = totalTime / this.stats.totalBatches;
    }

    // Update error rate (exponential moving average)
    const alpha = 0.1;
    this.stats.errorRate = alpha * errorRate + (1 - alpha) * this.stats.errorRate;

    // Calculate throughput
    this.stats.throughputPerSecond = processed / (timeTaken / 1000);
  }

  /**
   * Get processing statistics
   */
  getStats(): ProcessingStats {
    return { ...this.stats };
  }

  /**
   * Reset statistics
   */
  resetStats(): void {
    this.stats = {
      totalProcessed: 0,
      totalBatches: 0,
      averageBatchTime: 0,
      errorRate: 0,
      throughputPerSecond: 0,
    };
  }
}

/**
 * Simple semaphore implementation for concurrency control
 */
class Semaphore {
  private permits: number;
  private waitQueue: Array<() => void> = [];

  constructor(permits: number) {
    this.permits = permits;
  }

  async acquire(): Promise<void> {
    return new Promise((resolve) => {
      if (this.permits > 0) {
        this.permits--;
        resolve();
      } else {
        this.waitQueue.push(resolve);
      }
    });
  }

  release(): void {
    this.permits++;
    if (this.waitQueue.length > 0) {
      const resolve = this.waitQueue.shift()!;
      this.permits--;
      resolve();
    }
  }
}

export default ParallelProcessor;

File: core/parallel_processor.py
"""
Parallel Processing Engine for Todo Management System
"""
import asyncio
import time
import uuid
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed
from typing import List, Dict, Any, Callable, Optional, Union, TypeVar, Generic
from dataclasses import dataclass
from functools import wraps
import logging
from models.todo import Todo, BatchProcessingResult, TodoResponse
from sqlalchemy.orm import Session

logger = logging.getLogger(__name__)

T = TypeVar('T')
R = TypeVar('R')


@dataclass
class ProcessingBatch:
    """Represents a batch of items to be processed"""
    batch_id: str
    items: List[T]
    batch_size: int
    created_at: float

    def __post_init__(self):
        if not self.batch_id:
            self.batch_id = str(uuid.uuid4())
        if not self.created_at:
            self.created_at = time.time()


@dataclass
class ProcessingResult:
    """Result of processing a single item"""
    item_id: Any
    success: bool
    result: Any = None
    error: Optional[str] = None
    processing_time: float = 0.0


class ParallelProcessor(Generic[T, R]):
    """
    Advanced parallel processing engine for todo operations
    Supports both async and sync processing with intelligent batching
    """
    
    def __init__(
        self,
        max_workers: int = 4,
        batch_size: int = 10,
        max_concurrent_batches: int = 3,
        use_process_pool: bool = False,
        timeout_seconds: int = 300
    ):
        self.max_workers = max_workers
        self.batch_size = batch_size
        self.max_concurrent_batches = max_concurrent_batches
        self.use_process_pool = use_process_pool
        self.timeout_seconds = timeout_seconds
        
        # Initialize executors
        if use_process_pool:
            self.executor = ProcessPoolExecutor(max_workers=max_workers)
        else:
            self.executor = ThreadPoolExecutor(max_workers=max_workers)
        
        # Processing statistics
        self.stats = {
            'total_processed': 0,
            'total_batches': 0,
            'average_batch_time': 0.0,
            'error_rate': 0.0
        }
    
    def create_batches(self, items: List[T], custom_batch_size: Optional[int] = None) -> List[ProcessingBatch[T]]:
        """
        Intelligently partition items into batches for parallel processing
        """
        effective_batch_size = custom_batch_size or self.batch_size
        
        # Adaptive batch sizing based on item count
        if len(items) < effective_batch_size:
            effective_batch_size = max(1, len(items) // 2)
        elif len(items) > effective_batch_size * 10:
            # For large datasets, increase batch size slightly
            effective_batch_size = min(effective_batch_size * 2, len(items) // self.max_workers)
        
        batches = []
        for i in range(0, len(items), effective_batch_size):
            batch_items = items[i:i + effective_batch_size]
            batch = ProcessingBatch(
                batch_id=str(uuid.uuid4()),
                items=batch_items,
                batch_size=len(batch_items),
                created_at=time.time()
            )
            batches.append(batch)
        
        logger.info(f"Created {len(batches)} batches with average size {effective_batch_size}")
        return batches
    
    async def process_batch_async(
        self,
        batch: ProcessingBatch[T],
        processor_func: Callable[[T], R],
        **kwargs
    ) -> List[ProcessingResult]:
        """
        Process a single batch asynchronously
        """
        start_time = time.time()
        results = []
        
        try:
            # Create semaphore to limit concurrent operations within batch
            semaphore = asyncio.Semaphore(self.max_workers)
            
            async def process_item(item: T) -> ProcessingResult:
                async with semaphore:
                    item_start = time.time()
                    try:
                        # If processor_func is async
                        if asyncio.iscoroutinefunction(processor_func):
                            result = await processor_func(item, **kwargs)
                        else:
                            # Run sync function in thread pool
                            loop = asyncio.get_event_loop()
                            result = await loop.run_in_executor(
                                self.executor, processor_func, item
                            )
                        
                        return ProcessingResult(
                            item_id=getattr(item, 'id', str(item)),
                            success=True,
                            result=result,
                            processing_time=time.time() - item_start
                        )
                    except Exception as e:
                        logger.error(f"Error processing item {getattr(item, 'id', str(item))}: {e}")
                        return ProcessingResult(
                            item_id=getattr(item, 'id', str(item)),
                            success=False,
                            error=str(e),
                            processing_time=time.time() - item_start
                        )
            
            # Process all items in the batch concurrently
            tasks = [process_item(item) for item in batch.items]
            results = await asyncio.gather(*tasks, return_exceptions=True)
            
            # Handle any exceptions from gather
            processed_results = []
            for i, result in enumerate(results):
                if isinstance(result, Exception):
                    processed_results.append(ProcessingResult(
                        item_id=getattr(batch.items[i], 'id', str(batch.items[i])),
                        success=False,
                        error=str(result),
                        processing_time=0.0
                    ))
                else:
                    processed_results.append(result)
            
            processing_time = time.time() - start_time
            logger.info(f"Batch {batch.batch_id} processed in {processing_time:.2f}s")
            
            return processed_results
            
        except Exception as e:
            logger.error(f"Critical error processing batch {batch.batch_id}: {e}")
            # Return error results for all items in batch
            return [
                ProcessingResult(
                    item_id=getattr(item, 'id', str(item)),
                    success=False,
                    error=f"Batch processing failed: {str(e)}",
                    processing_time=0.0
                )
                for item in batch.items
            ]
    
    def process_batch_sync(
        self,
        batch: ProcessingBatch[T],
        processor_func: Callable[[T], R],
        **kwargs
    ) -> List[ProcessingResult]:
        """
        Process a single batch synchronously using thread/process pool
        """
        start_time = time.time()
        results = []
        
        def process_item_wrapper(item: T) -> ProcessingResult:
            item_start = time.time()
            try:
                result = processor_func(item, **kwargs)
                return ProcessingResult(
                    item_id=getattr(item, 'id', str(item)),
                    success=True,
                    result=result,
                    processing_time=time.time() - item_start
                )
            except Exception as e:
                logger.error(f"Error processing item {getattr(item, 'id', str(item))}: {e}")
                return ProcessingResult(
                    item_id=getattr(item, 'id', str(item)),
                    success=False,
                    error=str(e),
                    processing_time=time.time() - item_start
                )
        
        try:
            # Submit all items to executor
            future_to_item = {
                self.executor.submit(process_item_wrapper, item): item
                for item in batch.items
            }
            
            # Collect results as they complete
            for future in as_completed(future_to_item, timeout=self.timeout_seconds):
                try:
                    result = future.result()
                    results.append(result)
                except Exception as e:
                    item = future_to_item[future]
                    results.append(ProcessingResult(
                        item_id=getattr(item, 'id', str(item)),
                        success=False,
                        error=str(e),
                        processing_time=0.0
                    ))
            
            processing_time = time.time() - start_time
            logger.info(f"Batch {batch.batch_id} processed in {processing_time:.2f}s")
            
        except Exception as e:
            logger.error(f"Critical error processing batch {batch.batch_id}: {e}")
            # Return error results for remaining items
            for item in batch.items:
                if not any(r.item_id == getattr(item, 'id', str(item)) for r in results):
                    results.append(ProcessingResult(
                        item_id=getattr(item, 'id', str(item)),
                        success=False,
                        error=f"Batch processing failed: {str(e)}",
                        processing_time=0.0
                    ))
        
        return results
    
    async def process_parallel_async(
        self,
        items: List[T],
        processor_func: Callable[[T], R],
        custom_batch_size: Optional[int] = None,
        **kwargs
    ) -> BatchProcessingResult:
        """
        Process items in parallel using async approach
        """
        if not items:
            return BatchProcessingResult(
                total_processed=0,
                successful=0,
                failed=0,
                processing_time=0.0,
                batch_id=str(uuid.uuid4())
            )
        
        start_time = time.time()
        batches = self.create_batches(items, custom_batch_size)
        all_results = []
        
        # Process batches with concurrency limit
        semaphore = asyncio.Semaphore(self.max_concurrent_batches)
        
        async def process_batch_with_semaphore(batch: ProcessingBatch[T]) -> List[ProcessingResult]:
            async with semaphore:
                return await self.process_batch_async(batch, processor_func, **kwargs)
        
        # Process all batches
        batch_tasks = [process_batch_with_semaphore(batch) for batch in batches]
        batch_results = await asyncio.gather(*batch_tasks, return_exceptions=True)
        
        # Flatten results
        for batch_result in batch_results:
            if isinstance(batch_result, Exception):
                logger.error(f"Batch processing exception: {batch_result}")
                continue
            all_results.extend(batch_result)
        
        # Calculate statistics
        total_processed = len(all_results)
        successful = sum(1 for r in all_results if r.success)
        failed = total_processed - successful
        processing_time = time.time() - start_time
        
        # Update internal stats
        self._update_stats(total_processed, len(batches), processing_time, failed / total_processed if total_processed > 0 else 0)
        
        # Collect successful results
        successful_results = [r.result for r in all_results if r.success and r.result]
        
        # Collect errors
        errors = [
            {"item_id": r.item_id, "error": r.error}
            for r in all_results if not r.success
        ]
        
        return BatchProcessingResult(
            total_processed=total_processed,
            successful=successful,
            failed=failed,
            errors=errors,
            processing_time=processing_time,
            batch_id=str(uuid.uuid4()),
            results=successful_results if isinstance(successful_results[0], TodoResponse) if successful_results else []
        )
    
    def process_parallel_sync(
        self,
        items: List[T],
        processor_func: Callable[[T], R],
        custom_batch_size: Optional[int] = None,
        **kwargs
    ) -> BatchProcessingResult:
        """
        Process items in parallel using synchronous approach
        """
        if not items:
            return BatchProcessingResult(
                total_processed=0,
                successful=0,
                failed=0,
                processing_time=0.0,
                batch_id=str(uuid.uuid4())
            )
        
        start_time = time.time()
        batches = self.create_batches(items, custom_batch_size)
        all_results = []
        
        # Process batches sequentially (each batch processes items in parallel)
        for batch in batches:
            batch_results = self.process_batch_sync(batch, processor_func, **kwargs)
            all_results.extend(batch_results)
        
        # Calculate statistics
        total_processed = len(all_results)
        successful = sum(1 for r in all_results if r.success)
        failed = total_processed - successful
        processing_time = time.time() - start_time
        
        # Update internal stats
        self._update_stats(total_processed, len(batches), processing_time, failed / total_processed if total_processed > 0 else 0)
        
        # Collect successful results
        successful_results = [r.result for r in all_results if r.success and r.result]
        
        # Collect errors
        errors = [
            {"item_id": r.item_id, "error": r.error}
            for r in all_results if not r.success
        ]
        
        return BatchProcessingResult(
            total_processed=total_processed,
            successful=successful,
            failed=failed,
            errors=errors,
            processing_time=processing_time,
            batch_id=str(uuid.uuid4()),
            results=successful_results if successful_results and isinstance(successful_results[0], TodoResponse) else []
        )
    
    def _update_stats(self, processed: int, batches: int, time_taken: float, error_rate: float):
        """Update internal processing statistics"""
        self.stats['total_processed'] += processed
        self.stats['total_batches'] += batches
        
        # Update average batch time
        if self.stats['total_batches'] > 0:
            total_time = self.stats['average_batch_time'] * (self.stats['total_batches'] - batches) + time_taken
            self.stats['average_batch_time'] = total_time / self.stats['total_batches']
        
        # Update error rate (exponential moving average)
        alpha = 0.1
        self.stats['error_rate'] = alpha * error_rate + (1 - alpha) * self.stats['error_rate']
    
    def get_stats(self) -> Dict[str, Any]:
        """Get processing statistics"""
        return self.stats.copy()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.executor.shutdown(wait=True)
    
    async def __aenter__(self):
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        self.executor.shutdown(wait=True)
